### PR TITLE
Improved performance for epee serialization:

### PR DESCRIPTION
**Pull request by vtnerd**
- Removed copy of field names in binary deserialization
  - Removed copy of array values in binary deserialization
  - Removed copy of string values in json deserialization
  - Removed unhelpful allocation in json string value parsing
  - Removed copy of blob data on binary and json serialization

The changes to the binary deserialization and serialization **do** affect the p2p code, which includes the stuff in the key-value serialization code. The changes are fairly straightforward.

I also did some benchmarking:

`GET_BLOCKS_FAST` with 1000 blocks fetched and `pruned == true` (identical to wallet behavior).
Serialization within 1 ms. Deserialization ~4.9% speedup on ryzen 3900x and ~5.6% on i7-7Y75 (mobile). The raw numbers were 3 ms and 6 ms improvement respectively.

`BLOCK_HEADERS_RANGE` with 5000 block headers fetched (not necessarily standard wallet behavior but a larger JSON file). Serialization faster but within 1 ms. Deserialization ~37.7% speed on ryzen 3900x and ~42.6% on i7-7Y75 (mobile). The raw numbers were 17.9 ms and 34.5 ms improvement respectively. Possibly less/more in practice since it depends on the structures and data in use.

Each test was run 100 times before running in a loop 1000 times continuously. Only the last 1000 were used in measurements. The first 100 were an attempt to mitigate weird caching and turbo booting behavior to get consistency.

EDIT: These were run in Linux with Gcc 9.2.

### CODE REVIEW

File: contrib/epee/include/storages/parserse_base_utils.h [hunk 1]
```diff
@@ -157,7 +157,6 @@ namespace misc_utils
         while (fi != buf_end && ((lut[(uint8_t)*fi] & 32)) == 0)
           ++fi;
         val.assign(it, fi);
-        val.reserve(std::distance(star_end_string, buf_end));
```
Comment (moneromooo-monero) [inline contrib/epee/include/storages/parserse_base_utils.h]: Was this actually slower ?

File: contrib/epee/include/storages/portable_storage_from_json.h [hunk 2]
```diff
@@ -128,20 +128,20 @@ namespace epee
                   errno = 0;
                   int64_t nval = strtoll(val.data(), NULL, 10);
                   if (errno) throw std::runtime_error("Invalid number: " + std::string(val));
-                  stg.set_value(name, nval, current_section);              
+                  stg.set_value(name, int64_t(nval), current_section);
```
Comment (moneromooo-monero) [inline contrib/epee/include/storages/portable_storage_from_json.h]: Why is that needed since nval is already int64_t ?

Comment (vtnerd) [reply contrib/epee/include/storages/portable_storage_from_json.h]: The `static_assert` within `set_value` will complain about copying a value. Alternatively, I could change the check to ignore all `is_arithmetic` copies since the intent is to catch more complex copy functions.

File: contrib/epee/include/storages/parserse_base_utils.h (hunk 1 above)
Comment (vtnerd) [reply contrib/epee/include/storages/parserse_base_utils.h]: This was slower because typically the json string doesn't have any escapes so this usually creates an unnecessary allocation. This could arguably be moved above the assign, but then the json parser is typically requesting enough memory for the remainder of the json buffer which is a big over estimate (and seems to be harder on heap implementations). The `push_back` call will typically double the space needed each time the capacity is reached so the performance drop on escape handling is lessened.

Without this change, moving the json string values is actually slower due to the time spent searching within the heap (especially when the incoming json buffer is large). The difference is that the large allocation is not free'd after a copy (it is now moved and kept in use), and therefore the heap has to find another large chunk of memory instead of re-using a recently used large chunk. I wouldn't rely on this behavior because its fairly specific to the glibc I happen to be running during the tests - it is very repeatable on that box though. And it was a b**** to track down because reducing allocations + copies pretty much never increases wall clock time (except for when it does because heaps are a PITA).

File: contrib/epee/include/storages/portable_storage_from_json.h (hunk 2 above)
Comment (moneromooo-monero) [reply contrib/epee/include/storages/portable_storage_from_json.h]: I don't mind, I just wanted to understand. Does the cast make this expression a temporary, technically ?

File: contrib/epee/include/storages/parserse_base_utils.h (hunk 1 above)
Comment (moneromooo-monero) [reply contrib/epee/include/storages/parserse_base_utils.h]: Ah, I might have tested with worst case inputs :)

File: contrib/epee/include/storages/portable_storage_from_json.h (hunk 2 above)
Comment (vtnerd) [reply contrib/epee/include/storages/portable_storage_from_json.h]: Yes.

File: contrib/epee/include/storages/portable_storage.h [hunk 3]
```diff
@@ -59,7 +61,7 @@ namespace epee
       bool       get_value(const std::string& value_name, t_value& val, hsection hparent_section);
       bool       get_value(const std::string& value_name, storage_entry& val, hsection hparent_section);
       template<class t_value>
-      bool       set_value(const std::string& value_name, const t_value& target, hsection hparent_section);
+      bool       set_value(const std::string& value_name, t_value&& target, hsection hparent_section);
```
Comment (jagerman) [inline contrib/epee/include/storages/portable_storage.h]: This would be much cleaner as `t_value target` rather than `t_value&& target` so that the caller can decide whether to move or copy the argument.  Then you can keep all the `std::move()` changes for efficiency while dropping all the places where the caller has to use `int64_t(val)` and the like to make explicit copies at the call site.

Comment (jagerman) [reply contrib/epee/include/storages/portable_storage.h]: Oh never mind, you're taking this as universal reference.  I was misreading the diff.

File: contrib/epee/include/storages/portable_storage_from_json.h (hunk 2 above)
Comment (jagerman) [reply contrib/epee/include/storages/portable_storage_from_json.h]: It's an unnecessary change that makes parts of the code like this one worse.  It should be left up to the caller to decide whether passing an rvalue reference or making a copy is appropriate, and the `static_assert` should be a tool used to track down expensive copies *in a local branch* (or with some off-by-default local define guard), not something forced on all users of the interface.  The caller always has more information than the callee, and forcing the caller to make explicit copies everywhere they want to do so because someone somewhere *might* use a copy when they shouldn't is bad interface design.

Comment (xiphon) [reply contrib/epee/include/storages/portable_storage_from_json.h]: It actually makes things worse. The change utilizes intermediate CPU register for storing a value and then saving it to the stack instead of just using the stack.

Tested on MinGW 9.2.0 with default release compilation flags (i.e. -O2). 

Before:
```asm
	call	strtoll
	movq	%rax, 112(%rsp)
	call	*%r12
	movl	(%rax), %ecx
	testl	%ecx, %ecx
	jne	.L46392
	movq	72(%rsp), %rdx
	movq	%rbp, %r9
	leaq	112(%rsp), %r8
	movq	%rdi, %rcx
	call	_ZN4epee13serialization16portable_storage9set_valueIRxEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEOT_PNS0_7sectionE
```

After:
```asm
	call	strtoll
	movq	%rax, %r13
	call	*%r12
	movl	(%rax), %ecx
	testl	%ecx, %ecx
	jne	.L46293
	movq	72(%rsp), %rdx
	movq	%rbp, %r9
	leaq	112(%rsp), %r8
	movq	%rdi, %rcx
	movq	%r13, 112(%rsp)
	call	_ZN4epee13serialization16portable_storage9set_valueIxEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEOT_PNS0_7sectionE
```

Comment (vtnerd) [reply contrib/epee/include/storages/portable_storage_from_json.h]: > It's an unnecessary change that makes parts of the code like this one worse. It should be left up to the caller to decide whether passing an rvalue reference or making a copy is appropriate, and the static_assert should be a tool used to track down expensive copies in a local branch (or with some off-by-default local define guard), not something forced on all users of the interface. The caller always has more information than the callee, and forcing the caller to make explicit copies everywhere they want to do so because someone somewhere might use a copy when they shouldn't is bad interface design.

The callee (in `portable_storage`) **always** needs a copy of the value in this function. Its not obvious due to the way it is currently written (there is some outstanding inefficient map usage to be handled later). The function could take the argument by value and then move (as per recommendation by Scott Meyers), however in this situation my preference was to force copying to be explicit by the caller similar to Rust. Implicit copies can be kind of frustrating in this type of code (IMO at least).

----------------- 

@xiphon I don't see much of interest from the diassembly, except for odd output differences from mingw. The first version takes the output from `stroll` and immediately shoves it onto the stack for use in the `set_value` function which needs a reference to the value. In the second case, it shoves the output from `stroll` into `r13` and then takes the value from `r13` and shoves it onto the stack like the first version (which still needs a reference unfortunately due to C++ r-value rules).

There is a lesser-known library called `boost::call_traits` or something. It automatically drops references to arithmetic template arguments so when not-inlined, the compiler can pass the argument via register, etc. Not sure if it was updated for r-value stuff. But it would also mean not having to manually copy arithmetic values.

Comment (moneromooo-monero) [reply contrib/epee/include/storages/portable_storage_from_json.h]: The interesting part is what it does with std::string.

Comment (vtnerd) [reply contrib/epee/include/storages/portable_storage_from_json.h]: @xiphon I think you may have been pointing out that the extra "copy" here of the `int64_t` caused the output to change ... ? The output from gcc 9.2 target the Linux x86-64 doesn't do this extra work, the `rax` register is shoved onto the stack exactly once and everything is setup as expected. So I'm not sure what to do say about the mingw output ...

Comment (jagerman) [reply contrib/epee/include/storages/portable_storage_from_json.h]: > The function could take the argument by value and then move (as per recommendation by Scott Meyers), however in this situation my preference was to force copying to be explicit by the caller similar to Rust.

Can you elaborate on why you think it is better to import a foreign paradigm from another programming language than to listen to the advice of one of the most experienced, influential people in the C++ community?

Comment (xiphon) [reply contrib/epee/include/storages/portable_storage_from_json.h]: @vtnerd 
Same happens on Linux.

@moneromooo-monero 
> The interesting part is what it does with std::string.

It does an extra copy with `t_type(d)`.

Debian GCC 9.2, default release compilation flags (i.e. `-Ofast`).

Without `t_type(d)`, i.e. `return stg.set_value(pname, d, hparent_section);`:
```asm
_ZN4epee13serialization58kv_serialization_overloads_impl_is_base_serializable_typesILb1EE12kv_serializeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS0_16portable_storageEEEbRKT_RT0_NSE_8hsectionEPKc:
.LFB57876:
	.cfi_startproc
	.cfi_personality 0x9b,DW.ref.__gxx_personality_v0
	.cfi_lsda 0x1b,.LLSDA57876
	endbr64
	pushq	%r15
	.cfi_def_cfa_offset 16
	.cfi_offset 15, -16
	pushq	%r14
	.cfi_def_cfa_offset 24
	.cfi_offset 14, -24
	movq	%rdx, %r14
	movq	$-1, %rdx
	pushq	%r13
	.cfi_def_cfa_offset 32
	.cfi_offset 13, -32
	movq	%rdi, %r13
	pushq	%r12
	.cfi_def_cfa_offset 40
	.cfi_offset 12, -40
	movq	%rsi, %r12
	pushq	%rbp
	.cfi_def_cfa_offset 48
	.cfi_offset 6, -48
	movq	%rcx, %rbp
	pushq	%rbx
	.cfi_def_cfa_offset 56
	.cfi_offset 3, -56
	subq	$56, %rsp
	.cfi_def_cfa_offset 112
	movq	%fs:40, %rax
	movq	%rax, 40(%rsp)
	xorl	%eax, %eax
	leaq	16(%rsp), %rbx
	movq	%rbx, (%rsp)
	testq	%rcx, %rcx
	je	.L42377
	movq	%rcx, %rdi
	call	strlen@PLT
	leaq	0(%rbp,%rax), %rdx
.L42377:
	movq	%rsp, %r15
	movq	%rbp, %rsi
	movq	%r15, %rdi
.LEHB3709:
	call	_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag.constprop.0
.LEHE3709:
	movq	%r14, %rcx
	movq	%r13, %rdx
	movq	%r15, %rsi
	movq	%r12, %rdi
.LEHB3710:
	call	_ZN4epee13serialization16portable_storage9set_valueIRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbSA_OT_PNS0_7sectionE@PLT
```

With `t_type(d)`, i.e. `return stg.set_value(pname, t_type(d), hparent_section);`:
```asm
_ZN4epee13serialization58kv_serialization_overloads_impl_is_base_serializable_typesILb1EE12kv_serializeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS0_16portable_storageEEEbRKT_RT0_NSE_8hsectionEPKc:
.LFB57876:
	.cfi_startproc
	.cfi_personality 0x9b,DW.ref.__gxx_personality_v0
	.cfi_lsda 0x1b,.LLSDA57876
	endbr64
	pushq	%r15
	.cfi_def_cfa_offset 16
	.cfi_offset 15, -16
	pushq	%r14
	.cfi_def_cfa_offset 24
	.cfi_offset 14, -24
	pushq	%r13
	.cfi_def_cfa_offset 32
	.cfi_offset 13, -32
	pushq	%r12
	.cfi_def_cfa_offset 40
	.cfi_offset 12, -40
	movq	%rdx, %r12
	pushq	%rbp
	.cfi_def_cfa_offset 48
	.cfi_offset 6, -48
	movq	%rcx, %rbp
	pushq	%rbx
	.cfi_def_cfa_offset 56
	.cfi_offset 3, -56
	subq	$104, %rsp
	.cfi_def_cfa_offset 160
	movq	8(%rdi), %rdx
	movq	%rsi, 8(%rsp)
	movq	(%rdi), %rsi
	leaq	16(%rsp), %r15
	leaq	32(%rsp), %rbx
	movq	%r15, %rdi
	leaq	64(%rsp), %r13
	addq	%rsi, %rdx
	movq	%fs:40, %rax
	movq	%rax, 88(%rsp)
	xorl	%eax, %eax
	movq	%rbx, 16(%rsp)
.LEHB3615:
	call	_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag.constprop.0
.LEHE3615:
	movq	%r13, 48(%rsp)
	movq	$-1, %rdx
	testq	%rbp, %rbp
	je	.L41695
	movq	%rbp, %rdi
	call	strlen@PLT
	leaq	0(%rbp,%rax), %rdx
.L41695:
	leaq	48(%rsp), %r14
	movq	%rbp, %rsi
	movq	%r14, %rdi
.LEHB3616:
	call	_ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPKcEEvT_S8_St20forward_iterator_tag.constprop.0
.LEHE3616:
	movq	8(%rsp), %rdi
	movq	%r12, %rcx
	movq	%r15, %rdx
	movq	%r14, %rsi
.LEHB3617:
	call	_ZN4epee13serialization16portable_storage9set_valueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbRKS8_OT_PNS0_7sectionE@PLT
```

Comment (vtnerd) [reply contrib/epee/include/storages/portable_storage_from_json.h]: > Can you elaborate on why you think it is better to import a foreign paradigm from another programming language than to listen to the advice of one of the most experienced, influential people in the C++ community?

Yes, as already stated, I am trying to prevent implicit copies from occurring. It is extremely easy to miss at the call site, especially since the compiler will not "trap" or warn the user if the function parameters change from `const&` to by-value. So these checks partially helped me track down where copies occurred. So I think the checking could continue helping with maintenance. These are very frequently called functions within the codebase. And lastly, explicitly moving or requesting a copy at the call site is a trivial amount of typing and affects a tiny section of the codebase since all of this is hidden when adding fields to a p2p or rpc call.

-------------------

>  @vtnerd
> Same happens on Linux.

You've changed to an entirely differently function here -

This forcibly put a `std::string` copy into this function from another function. I still don't see your point - the copy occurs either way because `portable_storage` design requires ownership of the `std::string` from the struct that is going to be serialized. In other words, my changes add an explicit copy inside of this function instead of an implicit copy hidden within `set_value`. Any complaints about copying a `std::string` in this situation are primarily due to the existing design that I am hoping to slowly improve.

Comment (xiphon) [reply contrib/epee/include/storages/portable_storage_from_json.h]: Yes, i have to correct myself, it doesn't make things worse from memory management perspective, i.e. no extra copy should occur because 
> changes add an explicit copy inside of this function instead of an implicit copy hidden within `set_value`. Any complaints about copying a `std::string` in this situation are primarily due to the existing design that I am hoping to slowly improve.

Though every time i look at this code line it makes me wonder what is going on here and why do we now have an argument copy here.
